### PR TITLE
feat(search): implement `x-platform-adapter`

### PR DESCRIPTION
**Pull request by diegopf**
Implement endpoint adapters for platform API

[EX-5439](https://searchbroker.atlassian.net/browse/EX-5439)

### CODE REVIEW

File: packages/x-utils/src/__tests__/url.spec.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
+import { extractUrlParameters, getTaggingInfoFromUrl } from '../url';
+
+describe('url utils methods tests', () => {
```
Comment (LuisMartinez15) [inline packages/x-utils/src/__tests__/url.spec.ts:3]: Please wrap same function test cases in their own `describe`, so it is easier to read which tests belong to which function:

```
describe('extractUrlParameters', () => {
  ....
};

...

describe('getTaggingInfoFromUrl', () => {
  ....
};
```

File: packages/x-adapter/src/schemas/schemas.types.ts [hunk 2]
```diff
@@ -223,13 +223,28 @@ export type SubSchemaTransformer<Source, Target> = {
...
+ * @public
+ */
+export type SubSchema<Source, Target, Path extends ExtractPath<Source>> = ExtractType<
+  Source,
+  Path
+> extends (infer SourceArrayType)[]
+  ? Target extends (infer TargetArrayType)[]
+    ? Schema<SourceArrayType, TargetArrayType>
+    : never
+  : Target extends (infer TargetArrayType)[]
```
Comment (LuisMartinez15) [inline packages/x-adapter/src/schemas/schemas.types.ts:248]: I think you need to merge `main` branch into this one, as this is something that's already done there.

```suggestion
  : Target extends []
```

File: packages/x-adapter-platform/src/schemas/facet.schema.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  parentId: (_, $context) => $context?.parentId as string,
+  selected: () => false,
+  modelName: () => 'HierarchicalFilter',
+  children: {
+    $path: 'children',
+    $subSchema: facetMutableSchema as any
+  }
+};
+
+export const hierarchicalFilterMutableSchema = createMutableSchema(hierarchicalFilterSchema);
```
Comment (LuisMartinez15) [inline packages/x-adapter-platform/src/schemas/facet.schema.ts:46]: Can you move this `schema` to its own file? Same approach as with the other type of filters.

File: packages/x-adapter-platform/src/schemas/facet.schema.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  const name = facet.split('_')[0];
+  const hierarchicalFacet = ['categoryPaths'];
+  const numberRangeFacet = ['price'];
+  if (hierarchicalFacet.includes(name)) {
+    return hierarchicalFilterMutableSchema;
+  }
+  if (numberRangeFacet.includes(name)) {
+    return numberFilterMutableSchema;
+  }
+  return simpleMutableFilterSchema;
```
Comment (LuisMartinez15) [inline packages/x-adapter-platform/src/schemas/facet.schema.ts:83]: You can use the `getFacetType` helper here I guess

```suggestion
  const facetType = getFacetType(facet);
  if (facetType === 'HierarchicalFacet') {
    return hierarchicalFilterMutableSchema;
  }
  if (facetType === 'NumberRangeFacet') {
    return numberFilterMutableSchema;
  }
  return simpleMutableFilterSchema;
```

**Review by LuisMartinez15 — CHANGES_REQUESTED** [review]
Good job! 👏🏿 👏🏿 

Sadly, our platform `API` is not in a good state for testing nor analysing the response models, as some data is missing (`banners` and `promoted`).

File: packages/x-adapter-platform/src/schemas/facet.schema.ts (hunk 3 above)
Comment (diegopf) [reply packages/x-adapter-platform/src/schemas/facet.schema.ts:46]: Since `facetSchema` is depending on `hierarchicalFilterSchema` and viceversa I think makes sense to keep them together. Besides, I did not find an easy way to split them and make it work, to be honest 😅

File: packages/x-adapter-platform/src/schemas/next-query.schema.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
+import { createMutableSchema, Schema } from '@empathyco/x-adapter-next';
+import { NextQuery } from '@empathyco/x-types';
+import { PlatformNextQuery } from '../types';
```
Comment (LuisMartinez15) [inline packages/x-adapter-platform/src/schemas/next-query.schema.ts:3]:
```suggestion
import { PlatformNextQuery } from '../types/response.types';
```

File: packages/x-adapter-platform/src/endpoint-adapters/tagging.endpoint-adapter.ts [hunk 6]
```diff
@@ -0,0 +1,17 @@
+import {
+  beaconHttpClient,
+  endpointAdapterFactory,
+  EndpointAdapterOptions,
+  buildUrl
+} from '@empathyco/x-adapter-next';
+import { TaggingRequest } from '../types';
```
Comment (LuisMartinez15) [inline packages/x-adapter-platform/src/endpoint-adapters/tagging.endpoint-adapter.ts:7]:
```suggestion
import { TaggingRequest } from '../types/request.types';
```

File: packages/x-adapter-platform/src/endpoint-adapters/related-tags.endpoint-adapter.ts [hunk 7]
```diff
@@ -0,0 +1,15 @@
+import { endpointAdapterFactory, EndpointAdapterOptions } from '@empathyco/x-adapter-next';
+import { BaseRequest, RelatedTagsResponse } from '../types';
```
Comment (LuisMartinez15) [inline packages/x-adapter-platform/src/endpoint-adapters/related-tags.endpoint-adapter.ts:2]:
```suggestion
import { BaseRequest } from '../types/request.types';
import { RelatedTagsResponse } from '../types/response.types';
```

File: packages/x-adapter-platform/src/endpoint-adapters/sku-search.endpoint-adapter.ts [hunk 8]
```diff
@@ -0,0 +1,12 @@
+import { endpointAdapterFactory, EndpointAdapterOptions } from '@empathyco/x-adapter-next';
+import { BaseRequest, SkuSearchResponse } from '../types';
```
Comment (LuisMartinez15) [inline packages/x-adapter-platform/src/endpoint-adapters/sku-search.endpoint-adapter.ts:2]:
```suggestion
import { BaseRequest } from '../types/request.types';
import { SkuSearchResponse } from '../types/response.types';
```

File: packages/x-adapter-platform/src/endpoint-adapters/top-clicked.endpoint-adapter.ts [hunk 9]
```diff
@@ -0,0 +1,15 @@
+import { endpointAdapterFactory, EndpointAdapterOptions } from '@empathyco/x-adapter-next';
+import { BaseRequest, TopClickedResponse } from '../types';
```
Comment (LuisMartinez15) [inline packages/x-adapter-platform/src/endpoint-adapters/top-clicked.endpoint-adapter.ts:2]:
```suggestion
import { topClickedResponseMapper } from '../mappers/response/top-clicked-response.mapper';
import { baseRequestMapper } from '../mappers/request/base-request.mapper';
```

File: packages/x-adapter-platform/src/endpoint-adapters/next-queries.endpoint-adapter.ts [hunk 10]
```diff
@@ -0,0 +1,15 @@
+import { endpointAdapterFactory, EndpointAdapterOptions } from '@empathyco/x-adapter-next';
+import { BaseRequest, NextQueriesResponse } from '../types';
```
Comment (LuisMartinez15) [inline packages/x-adapter-platform/src/endpoint-adapters/next-queries.endpoint-adapter.ts:2]:
```suggestion
import { BaseRequest } from '../types/request.types';
import { NextQueriesResponse } from '../types/response.types';
```

File: packages/x-adapter-platform/src/__tests__/platform.adapter.spec.ts [hunk 11]
```diff
@@ -0,0 +1,467 @@
...
+import { Filter } from '@empathyco/x-types';
+import { platformAdapter } from '../platform.adapter';
+import {
+  BaseRequest,
+  PlatformEmpathizeResponse,
+  PlatformNextQueriesResponse,
+  PlatformRelatedTagsResponse,
+  PlatformSearchResponse,
+  TaggingRequest
+} from '../types';
```
Comment (LuisMartinez15) [inline packages/x-adapter-platform/src/__tests__/platform.adapter.spec.ts:11]:
```suggestion
import { BaseRequest, TaggingRequest } from '../types/request.types';
import {
  PlatformEmpathizeResponse,
  PlatformNextQueriesResponse,
  PlatformRelatedTagsResponse,
  PlatformSearchResponse
} from '../types/response.types';
```

File: packages/x-adapter-platform/src/mappers/__tests__/next-queries-resopnse.mapper.spec.ts [hunk 12]
```diff
@@ -0,0 +1,44 @@
+import { PlatformNextQueriesResponse } from '../../types';
```
Comment (tiborux) [inline packages/x-adapter-platform/src/mappers/__tests__/next-queries-resopnse.mapper.spec.ts:1]: Becareful with the name next-queries-resopnse.mapper.spec.ts

File: packages/x-adapter-platform/src/endpoint-adapters/empathize.endpoint-adapter.ts [hunk 13]
```diff
@@ -0,0 +1,13 @@
+import { endpointAdapterFactory, EndpointAdapterOptions } from '@empathyco/x-adapter-next';
+import { BaseRequest } from '../types/request.types';
+import { empathizeResponseMapper } from '../mappers/response/empathize-response.mapper';
+import { EmpathizeResponse } from '../types';
+import { baseRequestMapper } from '../mappers/request/base-request.mapper';
```
Comment (LuisMartinez15) [inline packages/x-adapter-platform/src/endpoint-adapters/empathize.endpoint-adapter.ts:5]:
```suggestion
import { baseRequestMapper } from '../mappers/request/base-request.mapper';
import { empathizeResponseMapper } from '../mappers/response/empathize-response.mapper';
import { BaseRequest } from '../types/request.types';
import { EmpathizeResponse } from '../types/response.types';
```

**Review by LuisMartinez15 — CHANGES_REQUESTED** [review]
Good job! I left "a lot of code suggestions" regarding the `imports`, as webstorm auto-imported stuff from barrels instead of specific files, so it is easier for you to fix them.

File: packages/x-adapter-platform/src/mappers/request/tagging-request.mapper.ts [hunk 14]
```diff
@@ -0,0 +1,10 @@
+import { createMutableSchema, Schema, schemaMapperFactory } from '@empathyco/x-adapter-next';
```
[the remark this answers is not in the document]
Comment (diegopf) [reply packages/x-adapter-platform/src/mappers/request/tagging-request.mapper.ts:1]: Nope, I've created it because we need it. 

The Tagging I think is quite special case since we only have to create the url coming from the `TaggingRequest`. This `TaggingRequest` is an object containing two fields: `params` and `url`. 

If we don't provide a request mapper then the default `identityMapper` will be used and the params and url fields from the tagging request will be mapped as query parameters. The main purpose of this `tagging-request.mapper` is to prevent that and make sure the `TaggingRequest` is correct. Otherwise we will have the right url with some query parameters appended like 
`...params=%5Bobject+Object%5D&url=https%3A%2F%2Fapi.staging.empathy.co%2Ftagging%2Fv1%2Ftrack%2Fempathy%2Fclick`

File: packages/x-adapter-platform/src/mappers/request/search-request.mapper.ts [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+export const searchRequestMutableSchema = createMutableSchema(searchRequestSchema);
+export const searchRequestMapper = schemaMapperFactory<SearchRequest, PlatformSearchRequest>(
+  searchRequestMutableSchema
+);
+
+/**
+ * Converts the filters to the shape the Platform's API is expecting.
+ *
+ * @param filters - The filters from our internal request.
+ *
```
Comment (tiborux) [inline packages/x-adapter-platform/src/mappers/request/search-request.mapper.ts:28]: Should we write an example here? I know that we have the test, but maybe for readability.

File: packages/x-adapter-platform/src/mappers/__tests__/response/search-response.mapper.spec.ts [hunk 16]
```diff
@@ -0,0 +1,385 @@
+import { searchResponseMapper } from '../../response/search-response.mapper';
+import { PlatformSearchResponse, SearchResponse } from '../../../types/response.types';
+import { platformSearchResponse } from '../../../__tests__/__fixtures__/platform-search.response';
+
+describe('search response mapper tests', () => {
+  it('should map the response', () => {
+    const response: SearchResponse = {
+      banners: [
```
Comment (tiborux) [inline packages/x-adapter-platform/src/mappers/__tests__/response/search-response.mapper.spec.ts:8]: Maybe we can move these responses to a file and mix the skuSearch and the search responses:

const skuSearch = {
  ....
  ....
  }
  
  const searchResponse = {
     banners: [],
     ...the rest of properties that doesnt exist on the skuSearch
     ...skuSearch
  }
  
  Im not sure if this is better and clean, mix both responses

File: packages/x-adapter-platform/src/mappers/__tests__/response/top-clicked-response.mapper.spec.ts [hunk 17]
```diff
@@ -0,0 +1,36 @@
+import { PlatformTopClickedResponse } from '../../../types/response.types';
+import { topClickedResponseMapper } from '../../response/top-clicked-response.mapper';
+// eslint-disable-next-line max-len
+import { platformTopClickedResponse } from '../../../__tests__/__fixtures__/platform-top-clicked.response';
+
+describe('top clicked response mapper tests', () => {
+  it('should map the top clicked', () => {
+    const response: PlatformTopClickedResponse =
```
Comment (tiborux) [inline packages/x-adapter-platform/src/mappers/__tests__/response/top-clicked-response.mapper.spec.ts:8]: Why we need to cast this? I can see that the `platformTopClickedResponse` is not PlatformTopClickedResponse type:

```
export const platformTopClickedResponse = {
  topclicked: {
    content: [
      {
        image: 'https://assets.empathy.co/images-demo/31335.jpg',
        gender: 'Men',
```

should we add the type there and fix the type? Because right now the PlatformTopClickedResponse has a PlatformResult as type for the content and it doesnt have some of the values that the response is returning, like color, gender, year...etc

File: packages/x-adapter-platform/src/types/response.types.ts [hunk 18]
```diff
@@ -0,0 +1,174 @@
...
+export interface PlatformResult {
+  name: string;
+  averageRating?: number;
+  id: string;
+  images?: string[];
+  image: string;
+  price: number;
+  sku?: string;
+  url: string;
+  tagging: PlatformTagging;
```
Comment (tiborux) [inline packages/x-adapter-platform/src/types/response.types.ts]: I suppose that we are not using the other values like year, color, popularity...etc

File: packages/x-adapter/src/types/adapter.types.ts [hunk 19]
```diff
@@ -69,7 +69,7 @@ export interface EndpointAdapterOptions<Request, Response> {
    * The {@link Mapper} in charge of adapting the request object so the backend API can
    * understand it.
    */
-  requestMapper?: Mapper<Request, Record<string, unknown>>;
+  requestMapper?: Mapper<Request, Record<string, any>>;
```
Comment (tiborux) [inline packages/x-adapter/src/types/adapter.types.ts:72]: why is this change needed?

File: packages/x-adapter/src/mappers/schema-mapper.factory.ts [hunk 20]
```diff
@@ -95,16 +100,35 @@ function applySubSchemaTransformer<Source, Target>(
...
+  }
+
+  const extendedContext: Dictionary<any> = {};
+  if ($context) {
+    Object.keys($context).forEach(key => {
+      if (['requestParameters', 'endpoint', 'to'].includes(key)) {
+        return;
+      }
+      extendedContext[key] = extractValue(source, $context[key] as ExtractPath<typeof source>);
+    });
```
Comment (LuisMartinez15) [inline packages/x-adapter/src/mappers/schema-mapper.factory.ts:115]: Using `Object.entries()` so we have direct access to the value too

```suggestion
    Object.entries($context).forEach(([key, value]) => {
      if (['requestParameters', 'endpoint', 'to'].includes(key)) {
        return;
      }
      extendedContext[key] = extractValue(source, value as ExtractPath<typeof source>);
    });
```

File: packages/x-adapter/src/mappers/schema-mapper.factory.ts [hunk 21]
```diff
@@ -95,16 +100,35 @@ function applySubSchemaTransformer<Source, Target>(
...
-      : mapSchema<typeof subSource, Target>(
-          subSource,
-          subSchema as Schema<typeof subSource, Target>,
-          context
-        );
+  if (!subSource) {
+    return;
+  }
+
+  const extendedContext: Dictionary<any> = {};
```
Comment (LuisMartinez15) [inline packages/x-adapter/src/mappers/schema-mapper.factory.ts:108]: You can safely remove the generic here as by default it is already `any`

```suggestion
  const extendedContext: Dictionary = {};
```

File: packages/x-adapter-platform/src/mappers/request/tagging-request.mapper.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (LuisMartinez15) [reply packages/x-adapter-platform/src/mappers/request/tagging-request.mapper.ts:1]: You're totally right. Very well explained too, thanks!

**Review by LuisMartinez15 — CHANGES_REQUESTED** [review]
one step closer

File: packages/x-adapter/src/mappers/schema-mapper.factory.ts [hunk 22]
```diff
@@ -95,16 +100,35 @@ function applySubSchemaTransformer<Source, Target>(
...
+
+  const extendedContext: Dictionary = {};
+  if ($context) {
+    Object.entries($context).forEach(([key, value]) => {
+      if (['requestParameters', 'endpoint', 'to'].includes(key)) {
+        return;
+      }
+      extendedContext[key] = extractValue(source, value as ExtractPath<typeof source>);
+    });
+  }
```
Comment (LuisMartinez15) [inline packages/x-adapter/src/mappers/schema-mapper.factory.ts:116]: You are probably hating me right now, but... can we refactor this to the following?

```suggestion
  const extendedContext = Object.entries($context ?? []).reduce(
    (extendedContext, [key, value]) => ({
      ...extendedContext,
      // eslint-disable-next-line @typescript-eslint/no-extra-parens
      ...(['requestParameters', 'endpoint', 'to'].includes(key)
        ? { key: extractValue(source, value as ExtractPath<typeof source>) }
        : {})
    }),
    {} as Dictionary
  );
```

**Review by LuisMartinez15 — CHANGES_REQUESTED** [review]
the last one! 😅
